src/incubator/TransitionAnimator/index.tsx

```diff
@@ -0,0 +1,140 @@
+/* eslint-disable react-hooks/exhaustive-deps */
+import React, {PropsWithChildren, useEffect, useCallback, useImperativeHandle} from 'react';
+import {View as RNView, LayoutChangeEvent} from 'react-native';
+import Animated, {useSharedValue, useAnimatedStyle, withSpring, withTiming, runOnJS} from 'react-native-reanimated';
+import View, {ViewProps} from '../../components/view';
+import {forwardRef, ForwardRefInjectedProps} from '../../commons/new';
+import useHiddenLocations, {HiddenLocation} from './useHiddenLocations';
+const AnimatedView = Animated.createAnimatedComponent(View);
+export {HiddenLocation as TransitionLocation};
+
+const DEFAULT_ANIMATION_VELOCITY = 300;
+const DEFAULT_ANIMATION_CONFIG = {velocity: DEFAULT_ANIMATION_VELOCITY, damping: 18, stiffness: 300, mass: 0.4};
+
+export type TransitionAnimationEndType = 'in' | 'out';
+
+export interface TransitionAnimatorProps extends ViewProps {
+  /**
+   * Callback to the animation end.
+   */
+  onAnimationEnd?: (animationType: TransitionAnimationEndType) => void;
+  /**
+   * If this is given there will be an enter animation from this location.
+   */
+  enterAnimationLocation?: HiddenLocation;
+  /**
+   * If this is given there will be an exit animation in this location.
+   */
+  exitAnimationLocation?: HiddenLocation;
+}
+
+type Props = PropsWithChildren<TransitionAnimatorProps> & ForwardRefInjectedProps;
+interface Statics {
+  animateOut: () => void;
+}
+
+const TransitionAnimator = (props: Props) => {
+  const {
+    onAnimationEnd,
+    enterAnimationLocation,
+    exitAnimationLocation,
+    forwardedRef,
+    style: propsStyle,
+    onLayout: propsOnLayout,
+    ...others
+  } = props;
+  const containerRef = React.createRef<RNView>();
+  const {onLayout: hiddenLocationOnLayout, hiddenLocations} = useHiddenLocations({containerRef});
+  const visible = useSharedValue<boolean>(!enterAnimationLocation);
+
+  const getLocation = (location?: HiddenLocation) => {
+    if (location === 'left' || location === 'right') {
+      return {x: hiddenLocations[location], y: 0};
+    } else if (location === 'top' || location === 'bottom') {
+      return {x: 0, y: hiddenLocations[location]};
+    } else {
+      return {x: 0, y: 0};
+    }
+  };
+
+  // Has to start at {0, 0} with {opacity: 0} so layout can be measured
+  const translateX = useSharedValue<number>(0);
+  const translateY = useSharedValue<number>(0);
+  const animatedStyle = useAnimatedStyle(() => {
+    return {
+      transform: [{translateX: translateX.value}, {translateY: translateY.value}],
+      // TODO: do we want to take the component's opacity here? - I think combining opacities is buggy
+      opacity: Number(visible.value)
+    };
+  }, []);
+
+  const animationEnded = useCallback((animationType: TransitionAnimationEndType) => {
+    onAnimationEnd?.(animationType);
+  },
+  [onAnimationEnd]);
+
+  const outAnimationEnd = useCallback((isFinished: boolean) => {
```

Comment:
Few things on the animate methods.. 
It took me some time to understand why are there so many animate methods. 

First, notice the order you have: 
outAnimationEnd, animate, animateOut, useImperativeHandler, inAnimationEnd, animateIn. 

I think we can order them in a more sensible order to make it more readable. 
the `useImperativeHandler` in the middle is not helping.. 

Second, I'd consider moving all of these methods into a separate hook. 

One last thing, some of the methods names were confusing to me.. 
For instance I read `inAnimationEnd` as a method that returns a boolean value.. 
Consider renaming some of the methods: enterAnimatedEnd, exitAnimationEnd, onEnter, onExist, etc..

Reply:
Moved things around and rename (some differently than suggested), let me know what you think.

Reply:
Of course it's the main functionality but it doesn't mean it can't be separated.. 
I think the big challenge we're facing with function components is their readability. 
Unlike classes they tend to become messy very easily and it's really difficult to understand the flow and logic of the code. 
With hooks the code become more organized and sectionalized. 
Usually, it's hard to tell what each useEffect does and to what part of the code it belong. 

If I'm using hooks, you can break it down like this which is way more readable.. 
```
const ... = useHiddenLocation()
const ... = useAnimatedTransition()
const ... = useBlaBla()

return <View>....</View>
```

Reply:
Ah, now I understand what you mean, I'll take another look.